### ISSUE TITLE
[VoiceOver] Unable to move to correct next item during continuous reading

### ISSUE THREAD

**Issue by ettore** [readium/r2-navigator-swift#197]
### What happened?

When VoiceOver is active and the user enables continuous reading (two-finger swipe down), the gesture to move to the next paragraph (single finger swipe right) may jump to the end of the chapter, and not to the next paragraph as in non-continuous reading mode.

The issue appears to be a Safari/WKWebView bug, because I have reproduced it there too. However, in normal Safari usage this manifests itself much less frequently (more on why later). 

This clip shows the problem in the R2 test app:

https://user-images.githubusercontent.com/289104/128434086-29692730-b32d-455b-a2f9-76281bcf0468.mov

And this shows the problem in Safari:

https://user-images.githubusercontent.com/289104/128432282-1dc40778-8cb4-4045-8201-a8aee4c3c8a0.mov

In both clips I start in single-item reading mode and then enable continuous reading (larger box is selected), and at that point a swipe to the next item brings you to the end of the chapter in the test app (the bottom left arrow is selected), and in Safari to the first word in italics. Both are after several paragraphs away from the reading point.

### Expected behavior

Moving to the next item should bring you to either the next paragraph or any other separating item (such as a differently formatted word, or the end of a sentence) within the current paragraph.

### How to reproduce?

- Launch R2 app and enable VoiceOver.
- Open Moby Dick or any other book.
- One finger single tap to begin play paragraph A.
- Two fingers swipe down to turn on continuous read: at this point a large block of text is selected, including (e.g.) paragraphs A, B, C.
- While Paragraph A is still being read, do a 2 fingers single tap to pause.
- Now swipe left to right to move to the item after the item that was being read (usually "item" means paragraph)
 
Observed: the reading position is changed to the end of the selected (large) block of text, i.e. end of paragraph C. In the R2 testapp clip above, this is the end of the _chapter_, because the continuous reading selection extended until the end of the entire chapter. In the Safari example the selection stopped at the first word that was formatted differently (with a `span`): if the Moby Dick contained a similar `span`, the same would have happened there.

Desired: the position should be moved to the next paragraph (B) at most.

This zip contains the first chapter of Little Women, which is in the public domain and it's what I used to reproduce the problem in Safari:
[littel-women-ch1.zip](https://github.com/readium/r2-navigator-swift/files/6942076/littel-women-ch1.zip)

### Environment

Mobile Safari
R2 testapp 2.2.0 -- The same happens with current top of develop

#### Readium versions

* `r2-shared-swift`: 2.1.0
* `r2-streamer-swift`: 2.1.0
* `r2-navigator-swift`: 2.1.0
* `r2-lcp-swift`: 2.1.0

#### Development environment

macOS: 10.15.7
platform: x86_64
carthage: 0.38.0
Xcode 12.4
Build version 12D4e

#### Testing device

* iOS version: 14.6
* Model: iPhone 11
* Is it an emulator? no -- but I also reproduced it in simulators

### Additional context

* Are you willing to fix the problem and contribute a pull request? Yes 

I identified the problem in how VoiceOver decides to group blocks of identically formatted text together when continuous reading is enabled. If paragraphs are visually pretty close to each other, the enclosing box may group the entire chapter together in one single block. This is what happens in the Moby Dick example.  

If we forcibly separate paragraphs by at least 1em via a custom CSS, VoiceOver will react by grouping text in smaller "boxes" that should coincide with what a typical user considers a paragraph. You can verify this in Safari by adding at the bottom of the `core.css` in the attached zip:
```
p {
	margin-bottom: 1em !important;
} 
```
Note that this solution is imperfect, because different EPUBS will use different tags to represent paragraphs. But in our testing this seems to work pretty reliably. We have yet to find a book where it does not work.

I have 2 solutions in mind but I am not sure which one (if any) is more appropriate. I will illustrate both.

**Comment by ettore** [comment]
SOLUTION 1: USER SETTINGS SOLUTION

ReadiumCSS includes the `--USER__paraSpacing` setting which is [currently not exposed by r2-shared](https://github.com/readium/r2-shared-swift/blob/develop/r2-shared-swift/Publication/User%20Settings/UserSettings.swift). If we added that to `ReadiumCSSReference` and `ReadiumCSSName` and added an `Incrementable` user property when VoiceOver is active, the problem is fixed. Something like
```
private func buildCssProperties() {
    [...]
    if UIAccessibility.isVoiceOverRunning {
        userProperties.addIncrementable(nValue: 0.5,
                                        min: 0,
                                        max: 2,
                                        step: 0.1,
                                        suffix: "em",
                                        reference: ReadiumCSSReference.paragraphMargins.rawValue,
                                        name: ReadiumCSSName.paragraphMargins.rawValue)
        }
}
```

PROS:
- integrates well with existing CSS logic since it's an existing setting
- does not alter the existing CSS logic
- can be exposed to the user to increase/decrease paragraph margins at will
- it is simple to configure 

CONS:
- uses a ReadiumCSS setting that may be destined for other purposes
- exposes a ReadiumCSS setting that perhaps was not meant to be exposed
- not extensible

**Comment by ettore** [comment]
SOLUTION 2: INJECT ADDITIONAL "ACCESSIBILITY" CSS

Add an "accessibility" CSS to ReadiumCSS and change [EPUBReflowableSpreadView::makeScripts](https://github.com/readium/r2-navigator-swift/blob/40bd1b3e7cf87c17c75f68895bcaf20f42de7fa8/r2-navigator-swift/EPUB/EPUBReflowableSpreadView.swift#L316) function to include it if VoiceOver is on. 

This is simple in the ["https" case](https://github.com/readium/r2-navigator-swift/blob/40bd1b3e7cf87c17c75f68895bcaf20f42de7fa8/r2-navigator-swift/EPUB/EPUBReflowableSpreadView.swift#L325-L339). We just have to modify [css-inline.js](https://github.com/readium/r2-navigator-swift/blob/develop/r2-navigator-swift/EPUB/Assets/Scripts/css-inline.js) and then inline the new CSS.

In the [non-https case](https://github.com/readium/r2-navigator-swift/blob/40bd1b3e7cf87c17c75f68895bcaf20f42de7fa8/r2-navigator-swift/EPUB/EPUBReflowableSpreadView.swift#L341-L347) I believe we'd have to modify the ReadiumCSS `css.js` script to conditionally include our new CSS at the bottom of the [injectCSS](https://github.com/readium/r2-navigator-swift/blob/40bd1b3e7cf87c17c75f68895bcaf20f42de7fa8/r2-navigator-swift/EPUB/Assets/Scripts/css.js#L5) function:
```
if (${readiumCSSAccessibilityEnabled}) {
    head.appendChild(createLink('ReadiumCSS-accessibility'));
}
```
And then set the flag if VoiceOver is on/off from swift.

PROS:
- extensible: if we need to tweak more CSS elements beside `p` we can do so

CONS:
- Less user configurable: this is buried into ReadiumCSS

I'm happy to work on a PR for either solution, or even an alternative solution if anybody knows it. I'm just not sure which one aligns better with R2 goals/strategy.

**Comment by mickael-menu** [comment]
There are two different things:

1. This is a bug in the `WKWebView`. Should Readium automatically adjust the paragraph spacing when VoiceOver is enabled to account for this? Maybe. It might be difficult to take all cases into account.

2. In any case, we need to provide a way for reading apps to fix this. I think both of your solutions are desirable actually. The second one is more complex and would be a good fit for the Injectable API (coming soon!) instead of implementing something  ad-hoc for the accessibility. In the meantime, I think we can implement your first solution and expose `--USER__paraSpacing`.

We're working on a generic layer for the presentation settings, but we also want for reading apps to be able to use any available Readium CSS properties. So no problem for exposing more properties until apps can do it with the API without modifying Readium.

**Comment by mickael-menu** [comment]
@ettore Let's mention this issue during the next dev call
